File: tech-studio-projects/financial-industry/crypto-trading-simulation/java/src/network/multicast/MulticastServer.cpp
#include "MulticastServer.h"
#include <arpa/inet.h>
#include <string.h>
#include <unistd.h>
#include <iostream>

MulticastServer::MulticastServer(int port) : sock(-1), running(false) {
    sock = socket(AF_INET, SOCK_DGRAM, 0);
    localAddr.sin_family = AF_INET;
    localAddr.sin_addr.s_addr = INADDR_ANY;
    localAddr.sin_port = htons(port);
    
    group.imr_multiaddr.s_addr = inet_addr("239.0.0.1"); // Multicast group
    group.imr_interface.s_addr = htonl(INADDR_ANY);
}

MulticastServer::~MulticastServer() {
    stop();
}

void MulticastServer::start() {
    if (bind(sock, (struct sockaddr*)&localAddr, sizeof(localAddr)) < 0) {
        std::cerr << "Bind failed" << std::endl;
        return;
    }
    setsockopt(sock, IPPROTO_IP, IP_ADD_MEMBERSHIP, (void*)&group, sizeof(group));
    
    running = true;
    char buffer[1024] = {0};

    std::cout << "Multicast server is listening..." << std::endl;
    while (running) {
        struct sockaddr_in fromAddr;
        socklen_t fromLen = sizeof(fromAddr);
        int bytesRead = recvfrom(sock, buffer, sizeof(buffer), 0, (struct sockaddr*)&fromAddr, &fromLen);
        if (bytesRead > 0) {
            std::cout << "Received: " << std::string(buffer, bytesRead) << std::endl;
        }
    }
}

void MulticastServer::stop() {
    if (running) {
        running = false;
        close(sock);
    }
}

File: tech-studio-projects/financial-industry/crypto-trading-simulation/java/src/network/multicast/MulticastServer.java
// MulticastServer.java
import java.io.IOException;
import java.net.DatagramPacket;
import java.net.InetAddress;
import java.net.MulticastSocket;



/**
 * public class Main {
    public static void main(String[] args) {
        MulticastServer server = new MulticastServer(12345); // Port number
        new Thread(server::start).start(); // Start the server in a new thread

        // Run for a while, then stop (for demonstration purposes)
        try {
            Thread.sleep(20000); // Let the server run for 20 seconds
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
        }

        server.stop(); // Stop the server
        System.out.println("Multicast server stopped.");
    }
}
 
*/
public class MulticastServer {

    private MulticastSocket multicastSocket;
    private InetAddress groupAddress;
    private boolean running;

    // Constructor
    public MulticastServer(int port) {
        try {
            multicastSocket = new MulticastSocket(port);
            groupAddress = InetAddress.getByName("239.0.0.1"); // Multicast group
            multicastSocket.joinGroup(groupAddress);
            running = false;
        } catch (IOException e) {
            System.err.println("Failed to create MulticastSocket: " + e.getMessage());
        }
    }

    // Method to start the server
    public void start() {
        running = true;
        System.out.println("Multicast server is listening...");

        // Buffer for receiving messages
        byte[] buffer = new byte[1024];

        while (running) {
            try {
                DatagramPacket packet = new DatagramPacket(buffer, buffer.length);
                multicastSocket.receive(packet); // Receive packet from multicast group
                String receivedMessage = new String(packet.getData(), 0, packet.getLength());
                System.out.println("Received: " + receivedMessage);
            } catch (IOException e) {
                if (running) {
                    System.err.println("Error receiving packet: " + e.getMessage());
                }
            }
        }
    }

    // Method to stop the server
    public void stop() {
        if (running) {
            running = false;
            try {
                multicastSocket.leaveGroup(groupAddress); // Leave the multicast group
                multicastSocket.close(); // Close the socket
            } catch (IOException e) {
                System.err.println("Failed to leave group or close socket: " + e.getMessage());
            }
        }
    }
}
